### PR TITLE
[libc++] Implements the new version header generator.

### PR DESCRIPTION
**Pull request by mordante**
The generator makes a few changes to the output
- removes the synopsis, it did not really show what was implemented correctly.
- the output now is clang-format clean.

This code uses the new FTM data structure. Since the contents of this structure are not up-to-date the code is only used in its tests.

### CODE REVIEW

File: libcxx/utils/generate_feature_test_macro_components.py [hunk 1]
```diff
@@ -2060,12 +2109,106 @@ def implemented_ftms(self) -> Dict[str, Dict[str, Any]]:
 
         return get_ftms(self.__data, self.std_dialects, True)
 
+    @functools.cached_property
+    def ftm_meta_data(self) -> Dict[str, Dict[str, Any]]:
```
Comment (ldionne) [inline libcxx/utils/generate_feature_test_macro_components.py:2113]:
```suggestion
    def ftm_metadata(self) -> Dict[str, Dict[str, Any]]:
```

Metadata is generally spelled as a single word, I think?

File: libcxx/utils/generate_feature_test_macro_components.py [hunk 2]
```diff
@@ -2060,12 +2109,106 @@ def implemented_ftms(self) -> Dict[str, Dict[str, Any]]:
...
+        """Returns the meta data of the FTMs defined in the Standard.
+
+        The meta data does not depend on the C++ dialect used.
+        The result is a dict with the following contents:
+        - key: Name of the feature test macro.
+        - value: A dict with the following content:
+          * headers: The list of headers that should provide the FTM
+          * test_suite_guard: The condtion for testing the FTM in the test suite.
+          * test_suite_guard: The condtion for testing the FTM in the version header.
```
Comment (ldionne) [inline libcxx/utils/generate_feature_test_macro_components.py:2122]:
```suggestion
          * test_suite_guard: The condition for testing the FTM in the test suite.
          * test_suite_guard: The condition for testing the FTM in the version header.
```

**Review by ldionne — APPROVED** [review]
LGTM w/ a few comments.

File: libcxx/test/libcxx/feature_test_macro/version_header.sh.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+#  define __cpp_lib_variant 202102L
+#endif // _LIBCPP_STD_VER >= 17
+
+#if _LIBCPP_STD_VER >= 20
+#  if !defined(_LIBCPP_HAS_NO_THREADS) && _LIBCPP_AVAILABILITY_HAS_SYNC
+#    define __cpp_lib_barrier 201907L
+#  endif
+// define __cpp_lib_format 202110L
+#  undef __cpp_lib_variant
+#  define __cpp_lib_variant 202106L
```
Comment (cpplearner) [inline libcxx/test/libcxx/feature_test_macro/version_header.sh.py:54]: The output seems incorrect: this value is marked as `"implemented": false` in `test_data.json`, and thus it should not be defined in the generated output.

https://github.com/llvm/llvm-project/blob/a149e3919a5de7a43d933aa99c1cb6697e597d39/libcxx/test/libcxx/feature_test_macro/test_data.json#L136-L142

(I noticed this while working on [an up-to-date FTM data file](https://github.com/cpplearner/llvm-project/blob/ftm-data/libcxx/utils/feature_test_macro_data.json).

Comment (ldionne) [reply libcxx/test/libcxx/feature_test_macro/version_header.sh.py:54]: Thanks for noticing! This PR should fix it: https://github.com/llvm/llvm-project/pull/108843

**Pull request merged**